embed test_certs, remove reliance of tests on cwd

tbg opened the pull request:
the test certs are now embedded in package resource.
refactored the tls testing code so that it just pulls the asset.

as a nice side effect, the `make testbuild` task got easier since it doesn't
have to move the tests binaries to the correct location any more. the deploy
image creation script simplifies accordingly.

tbg committed:
embed test_certs, remove reliance of tests on cwd

the test certs are now embedded in package resource.
refactored the tls testing code so that it just pulls the asset.

as a nice side effect, the `make testbuild` task got easier since it doesn't
have to move the tests binaries to the correct location any more. the deploy
image creation script simplifies accordingly.

bdarnell commented on resource/resource.go:
```diff
@@ -22,6 +22,6 @@ package resource
 // avoid embedding any files (but instead passes them through from disk). This
 // avoids having to recompile before testing changes.
-//go:generate go-bindata -pkg resource -mode 0644 -modtime 1400000000 -o ./embedded.go ./ui/...
+//go:generate go-bindata -pkg resource -mode 0644 -modtime 1400000000 -o ./embedded.go ./ui/... ./test_certs/...
```
I'd like to embed the certs in a separate package that is only imported from tests. I don't want release binaries to contain the test certs since it would be a bad idea to ever actually use them.

tschottdorf replied:
Aren't they optimized away in the main build? But of course I can just move them over.

bdarnell commented on rpc/tls.go:
```diff
@@ -99,10 +110,22 @@ func LoadInsecureTLSConfig() *TLSConfig {
 // TODO Maybe instead of returning err, take a testing.T?  And move to tls_test?
-func LoadTestTLSConfig(projectRoot string) (*TLSConfig, error) {
-	return LoadTLSConfig(path.Join(projectRoot, "resource", "test_certs"))
+func LoadTestTLSConfig(certDir string) (*TLSConfig, error) {
```
This function should also move to the package where the testing certs live (cockroach/rpc/rpctest?)

bdarnell commented:
LGTM aside from the package moves.

cockroach-team commented:
Rather then doing this, you might use:
 _, file, _ :=runtime.Caller(0)
dir := filepath.Split(file)
filepath.Join(dir, "../test_certs/")

This actually works in tests.
-Daniel



On Thu, Mar 12, 2015 at 12:14 PM Ben Darnell <notifications@github.com>
wrote:

> LGTM aside from the package moves.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/416#issuecomment-78575644>.
>
> --
> You received this message because you are subscribed to the Google Groups
> "Cockroach DB" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to cockroach-db+unsubscribe@googlegroups.com.
> For more options, visit https://groups.google.com/d/optout.
>

tschottdorf commented:
@Daniel: Thanks, but it's still an external dependency. I'd rather have the `rpc.test` binary work just like that.

tbg committed:
embed test certs in package rpc/rpctest

LoadTestTLSConfig remained in rpc/tls.go since rpctest
imports rpc, leading to an import cycle.

tschottdorf commented:
Separated the certs out into package `rpctest`.
@bdarnell PTAL.
Wasn't able to move the `LoadTestTLSConfig` method to `rpctest` due to an import cycle (could factor out the use of RPC there, but then we'd need a helper in each test - not worth it). Moving to `tls_test.go` not possible since `server` tests want to use it (and apparently don't get other pkg's test code). Is there an elegant way out?

cockroach-team commented:
Have you considered moving it to the server package? I think that worked
for me. (I was working on a change but got busy with baby and work).

On Thu, Mar 12, 2015 at 12:43 PM Tobias Schottdorf <notifications@github.com>
wrote:

> Separated the certs out into package rpctest.
> @bdarnell <https://github.com/bdarnell> PTAL.
> Wasn't able to move the LoadTestTLSConfig method to rpctest due to an
> import cycle (could factor out the use of RPC there, but then we'd need a
> helper in each test - not worth it). Moving to tls_test.go not possible
> since server tests want to use it (and apparently don't get other pkg's
> test code). Is there an elegant way out?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/416#issuecomment-78588943>.
>
> --
> You received this message because you are subscribed to the Google Groups
> "Cockroach DB" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to cockroach-db+unsubscribe@googlegroups.com.
> For more options, visit https://groups.google.com/d/optout.
>

cockroach-team commented:
No, I added it to the util package. Sorry.

On Thu, Mar 12, 2015 at 12:53 PM Daniel Theophanes <kardianos@gmail.com>
wrote:

> Have you considered moving it to the server package? I think that worked
> for me. (I was working on a change but got busy with baby and work).
>
> On Thu, Mar 12, 2015 at 12:43 PM Tobias Schottdorf <
> notifications@github.com> wrote:
>
>> Separated the certs out into package rpctest.
>> @bdarnell <https://github.com/bdarnell> PTAL.
>> Wasn't able to move the LoadTestTLSConfig method to rpctest due to an
>> import cycle (could factor out the use of RPC there, but then we'd need a
>> helper in each test - not worth it). Moving to tls_test.go not possible
>> since server tests want to use it (and apparently don't get other pkg's
>> test code). Is there an elegant way out?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/cockroachdb/cockroach/pull/416#issuecomment-78588943>
>> .
>>
>> --
>> You received this message because you are subscribed to the Google Groups
>> "Cockroach DB" group.
>> To unsubscribe from this group and stop receiving emails from it, send an
>> email to cockroach-db+unsubscribe@googlegroups.com.
>> For more options, visit https://groups.google.com/d/optout.
>>
>

bdarnell commented:
LGTM

One way out would be to make LoadTestTLSConfig return the the three PEM byte arrays (or a struct that wraps them) instead of calling LoadTLSConfig for you.
